### PR TITLE
Feature/openstack blockstorage volume v2 data source

### PR DESCRIPTION
**Pull request by Shonei**

### CODE REVIEW

File: website/docs/d/blockstorage_volume_v2.html.markdown [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+# openstack\_blockstorage\_volume\_v2
+
+Use this data source to get information about an existing volume.
+
+## Example Usage
+
+```hcl
+data "openstack_blockstorage_volume_v2" "volume_1" {
+  name        = "volume_1"
```
Comment (jeffa5) [inline website/docs/d/blockstorage_volume_v2.html.markdown]: Remove the extra space here?

File: openstack/data_source_openstack_blockstorage_volume_v2.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
+package openstack
+
+import (
+	"fmt"
+	"log"
+
```
Comment (foamdino) [inline openstack/data_source_openstack_blockstorage_volume_v2.go:6]: remove unused whitespace / extra lines

File: openstack/data_source_openstack_blockstorage_volume_v2.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+	if err != nil {
+		return fmt.Errorf("Unable to query openstack_blockstorage_volume_v2: %s", err)
+	}
+
+	allVolumes, err := volumes.ExtractVolumes(allPages)
+	if err != nil {
+		return fmt.Errorf("Unable to retrieve openstack_blockstorage_volume_v2: %s", err)
+	}
+
+	if len(allVolumes) > 1 {
```
Comment (foamdino) [inline openstack/data_source_openstack_blockstorage_volume_v2.go:90]: do we want to support fetching all matches and failing when we expect a single match?

File: openstack/data_source_openstack_blockstorage_volume_v2.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+	d.Set("bootable", volume.Bootable)
+	d.Set("volume_type", volume.VolumeType)
+	d.Set("size", volume.Size)
+	d.Set("source_volume_id", volume.SourceVolID)
+
+	if err := d.Set("metadata", volume.Metadata); err != nil {
+		log.Printf("[DEBUG] Unable to set metadata for volume %s: %s", volume.ID, err)
+	}
+
+	return nil
```
Comment (foamdino) [inline openstack/data_source_openstack_blockstorage_volume_v2.go:114]: should we be returning nil here?

File: openstack/data_source_openstack_blockstorage_volume_v2.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+	d.SetId(volume.ID)
+	d.Set("name", volume.Name)
+	d.Set("status", volume.Status)
+	d.Set("bootable", volume.Bootable)
+	d.Set("volume_type", volume.VolumeType)
+	d.Set("size", volume.Size)
+	d.Set("source_volume_id", volume.SourceVolID)
+
+	if err := d.Set("metadata", volume.Metadata); err != nil {
+		log.Printf("[DEBUG] Unable to set metadata for volume %s: %s", volume.ID, err)
```
Comment (foamdino) [inline openstack/data_source_openstack_blockstorage_volume_v2.go:111]: should we hardcode `[DEBUG]` - surely that would be set by the logging framework

File: openstack/provider.go [hunk 6]
```diff
@@ -278,6 +278,7 @@ func Provider() terraform.ResourceProvider {
 			"openstack_sharedfilesystem_snapshot_v2":             dataSourceSharedFilesystemSnapshotV2(),
 			"openstack_keymanager_secret_v1":                     dataSourceKeyManagerSecretV1(),
 			"openstack_keymanager_container_v1":                  dataSourceKeyManagerContainerV1(),
+			"openstack_blockstorage_volume_v2":                   dataSourceBlockStorageVolumeV2(),
```
Comment (foamdino) [inline openstack/provider.go]: move this to just under the other `openstack_blockstorage_` data sources

**Review by foamdino — CHANGES_REQUESTED** [review]
Some changes required and some thoughts around how many results to return

File: openstack/data_source_openstack_blockstorage_volume_v2.go (hunk 2 above)
Comment (jeffa5) [reply openstack/data_source_openstack_blockstorage_volume_v2.go:6]: They are a part of `go fmt` I believe. Separates `std`, `external` and `local` imports.

File: openstack/data_source_openstack_blockstorage_volume_v2.go (hunk 5 above)
Comment (Shonei) [reply openstack/data_source_openstack_blockstorage_volume_v2.go:111]: That is how all logging is handled in the provider. Only difference is some use `[INFO]`

File: openstack/data_source_openstack_blockstorage_volume_v2.go (hunk 4 above)
Comment (Shonei) [reply openstack/data_source_openstack_blockstorage_volume_v2.go:114]: I think we should. Don't think setting the `metadata` should fail the run. Other data sources follow the same pattern as well.

File: openstack/data_source_openstack_blockstorage_volume_v2.go (hunk 3 above)
Comment (Shonei) [reply openstack/data_source_openstack_blockstorage_volume_v2.go:90]: I think we should only return a single result. If it is an array you can match someone else volume and I am not sure how you will be able to use an array of volumes. Also in the schema I am not sure you can return an array of objects but we can have some of the return parameters be an array.

File: openstack/data_source_openstack_blockstorage_volume_v2.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+		return fmt.Errorf("Unable to query openstack_blockstorage_volume_v2: %s", err)
+	}
+
+	allVolumes, err := volumes.ExtractVolumes(allPages)
+	if err != nil {
+		return fmt.Errorf("Unable to retrieve openstack_blockstorage_volume_v2: %s", err)
+	}
+
+	if len(allVolumes) > 1 {
+		return fmt.Errorf("Your openstack_blockstorage_volume_v2 query returned multiple results for your query.")
```
Comment (jeffa5) [inline openstack/data_source_openstack_blockstorage_volume_v2.go]: Reword, doesn't make sense

**Review by jeffa5 — CHANGES_REQUESTED** [review]

File: website/openstack.erb [hunk 8]
```diff
@@ -22,6 +22,9 @@
             <li<%= sidebar_current("docs-openstack-datasource-blockstorage-snapshot-v3") %>>
               <a href="/docs/providers/openstack/d/blockstorage_snapshot_v3.html">openstack_blockstorage_snapshot_v3</a>
             </li>
+            <li<%= sidebar_current("docs-openstack-datasource-blockstorage-volume-v2") %>>
+              <a href="/docs/providers/openstack/d/blockstorage_volume_v2.html">openstack_blockstorage_snapshot_v3</a>
```
Comment (dimitarvdimitrov) [inline website/openstack.erb]: `openstack_blockstorage_snapshot_v3`? 😮

File: website/docs/d/blockstorage_volume_v2.html.markdown [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+`id` is set to the ID of the found volume. In addition, the following attributes
+are exported:
+
+* `region` - See Argument Reference above.
+* `name` - See Argument Reference above.
+* `status` - See Argument Reference above.
+* `metadata` - See Argument Reference above.
+* `volume_type` - The type of the volume.
+* `bootable` - Indicates if the volume is bootable..
+* `size` - The size of the volume.
```
Comment (dimitarvdimitrov) [inline website/docs/d/blockstorage_volume_v2.html.markdown]: In bytes?

File: website/docs/d/blockstorage_volume_v2.html.markdown [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+`id` is set to the ID of the found volume. In addition, the following attributes
+are exported:
+
+* `region` - See Argument Reference above.
+* `name` - See Argument Reference above.
+* `status` - See Argument Reference above.
+* `metadata` - See Argument Reference above.
+* `volume_type` - The type of the volume.
+* `bootable` - Indicates if the volume is bootable..
```
Comment (dimitarvdimitrov) [inline website/docs/d/blockstorage_volume_v2.html.markdown]: Redundant `.`

File: openstack/data_source_openstack_blockstorage_volume_v2.go [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+func dataSourceBlockStorageVolumeV2Read(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	client, err := config.blockStorageV2Client(GetRegion(d, config))
+	if err != nil {
+		return fmt.Errorf("Error creating OpenStack block storage client: %s", err)
+	}
+	m := make(map[string]string)
+	for key, val := range d.Get("metadata").(map[string]interface{}) {
```
Comment (dimitarvdimitrov) [inline openstack/data_source_openstack_blockstorage_volume_v2.go]: You can use `expandToMapStringString()`

**Review by dimitarvdimitrov — CHANGES_REQUESTED** [review]

**Review by dimitarvdimitrov — APPROVED** [review]

**Review by foamdino — APPROVED** [review]
So long as it follows the patterns / best practices of the rest of the go community etc then I'm fine with it

**Review by jeffa5 — APPROVED** [review]

**Pull request merged**